src/rgw/rgw_rest_bucket_logging.cc

```diff
@@ -192,6 +192,41 @@ class RGWPutBucketLoggingOp : public RGWDefaultResponseOp {
       return;
     }
     auto& target_attrs = target_bucket->get_attrs();
+    // verify target permissions for bucket logging
+    // this is implementing the policy based permission granting from:
+    // https://docs.aws.amazon.com/AmazonS3/latest/userguide/enable-server-access-logging.html#grant-log-delivery-permissions-general
+    if (const auto policy = get_iam_policy_from_attr(s->cct, target_attrs, target_bucket_id.tenant); policy) {
+      ldpp_dout(this, 20) << "INFO: logging bucket '" << target_bucket_id <<
+        "' policy: " << *policy << dendl;
+      rgw::auth::ServiceIdentity ident(rgw::bucketlogging::service_principal);
+      // Note: the "eval()" call ignores mismatches in the principlal
+      // so, this is explicitly checked here
+      if (policy->eval_principal(s->env, ident) == rgw::IAM::Effect::Deny) {
+        ldpp_dout(this, 1) << "ERROR: target bucket '" << target_bucket_id <<
+          "' must have a bucket policy that with logging service principal '" <<
+          rgw::bucketlogging::service_principal << "'" << dendl;
+        op_ret = -EACCES;
+        return;
+      }
+      const auto target_resource_arn = rgw::ARN(target_bucket_id, configuration.target_prefix);
+      const auto source_bucket_arn = rgw::ARN(src_bucket_id);
+      // Note: ARNLike operator, often used in the SourceArn currently not supported
+      s->env.emplace("aws:SourceArn", source_bucket_arn.to_string());
+      s->env.emplace("aws:SourceAccount", s->account_name);
+      // Note: resource ARN mismatch do not deny
+      if (policy->eval(s->env, ident, rgw::IAM::s3PutObject, target_resource_arn) == rgw::IAM::Effect::Deny) {
+        ldpp_dout(this, 1) << "ERROR: target bucket '" << target_bucket_id <<
+          "' must have a bucket policy that allows logging service principal to put objects in the following resource: ARN '" <<
+          target_resource_arn.to_string() << "' from source bucket ARN'" << source_bucket_arn.to_string() <<
+          "' and account '" << s->account_name << "'" <<  dendl;
+        op_ret = -EACCES;
```

Comment:
in the standup, i think we discussed moving these permission checks to the write of each log entry, so this comment may not apply to the final version

but request permission checks should always happen in the `verify_permission()` override, because of how the admin/system user overrides work in https://github.com/ceph/ceph/blob/e56665c/src/rgw/rgw_process.cc#L226-L236

if `execute()` returns `EACCES`, admin users would not have superuser access for PutBucketLogging like they're supposed to

Reply:
* i thought of adding the check to both places? do you think we should keep that only when writing the log (also related to: https://github.com/ceph/ceph/pull/62284#discussion_r1998859909) ?
* addit the check in `verify_permission()` is possible in the case of `PutbucketLogging` op. although it would mean that we do all the heavy listing there
* but i don't think it is possible in the log writing case.